src/learn/gensfen.cpp

```diff
@@ -316,19 +318,86 @@ namespace Learner
                 // Discard stuff before write_minply is reached
                 // because it can harm training due to overfitting.
                 // Initial positions would be too common.
-                if (ply >= params.write_minply && !was_seen_before(pos))
+                if (ply >= params.write_minply)
                 {
                     packed_sfens.emplace_back(PackedSfenValue());
 
                     auto& psv = packed_sfens.back();
 
-                    // Here we only write the position data.
-                    // Result is added after the whole game is done.
-                    pos.sfen_pack(psv.sfen);
+                    if (params.ensure_quiet)
+                    {
+                        auto [qsearch_value, qsearch_pv] = Search::qsearch(pos);
+                        if (qsearch_pv.empty())
+                        {
+                            // Here we only write the position data.
+                            // Result is added after the whole game is done.
+                            pos.sfen_pack(psv.sfen);
+
+                            // Already a quiet position
+                            psv.score = search_value;
+                            psv.move = search_pv[0];
+                            psv.gamePly = ply;
+                        }
+                        else
+                        {
+                            // Navigate to a quiet
+                            int old_ply = ply;
+                            for (auto m : qsearch_pv)
+                            {
+                                pos.do_move(m, states[ply++]);
+                            }
+
+                            if (was_seen_before(pos))
+                            {
+                                // Just skip the move.
+                                packed_sfens.pop_back();
+                            }
+                            else
+                            {
+                                // Reevaluate
+                                auto [quiet_search_value, quiet_search_pv] = Search::search(pos, depth, 1, params.nodes);
```

Comment:
I don't understand why we need to search again.  Could you please explain about it?

Reply:
Because if the position after qsearch changed then the previous eval may be incorrect.

Reply:
In my understanding, "incorrect" here means that `quiet_search_value` is more accurate than `search_value`.  Because `quiet_search_value` is calculated with the search against an ancestor node of a quite position but `search_value` is calculated with the search against the quite position.  Is my understanding right?

Reply:
by incorrect I mean illegal. quiet_search_value is the value of the position after PV from qsearch was applied to the position. The searches are on two different positions and one cannot apply the move from the former to the psv of the latter.

Reply:
Yes. One search is done before qsearch, one after. They may have different evals and want to keep the one from the position saved in the psv. Note that this behaviour differs from `learn`, where we assume that after qsearch the actual eval doesn't deviate much from the stored value.

Reply:
I see.  It sounds good.  Thank you.